sycl-fusion/passes/target/TargetFusionInfo.cpp

```diff
@@ -0,0 +1,375 @@
+//==---------------------- TargetFusionInfo.cpp ----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TargetFusionInfo.h"
+
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/InstrTypes.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsNVPTX.h"
+#include "llvm/TargetParser/Triple.h"
+
+namespace llvm {
+class TargetFusionInfoImpl {
+
+public:
+  explicit TargetFusionInfoImpl(llvm::Module *Mod) : LLVMMod{Mod} {};
+
+  virtual ~TargetFusionInfoImpl() = default;
+
+  virtual void notifyFunctionsDelete(
+      [[maybe_unused]] llvm::ArrayRef<Function *> Funcs) const {}
+
+  virtual void addKernelFunction([[maybe_unused]] Function *KernelFunc) const {}
+
+  virtual void postProcessKernel([[maybe_unused]] Function *KernelFunc) const {}
+
+  virtual ArrayRef<StringRef> getKernelMetadataKeys() const { return {}; }
+
+  virtual ArrayRef<StringRef> getUniformKernelAttributes() const { return {}; }
+
+  virtual void createBarrierCall(IRBuilderBase &Builder,
+                                 int BarrierFlags) const = 0;
+
+  virtual unsigned getPrivateAddressSpace() const = 0;
+
+  virtual unsigned getLocalAddressSpace() const = 0;
+
+  virtual void
+  updateAddressSpaceMetadata([[maybe_unused]] Function *KernelFunc,
+                             [[maybe_unused]] ArrayRef<size_t> LocalSize,
+                             [[maybe_unused]] unsigned AddressSpace) const {}
+
+protected:
+  llvm::Module *LLVMMod;
+};
+
+//
```

Comment:
Place `SPIRVTargetFusionInfo` and `NVPTXTargetFusionInfo` declarations in an unnamed namespace for internal linkage.